hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -14,8 +32,296 @@ import HIE.Bios as BIOS
 -- of the project that may or may not be accurate.
 findLocalCradle :: FilePath -> IO Cradle
 findLocalCradle fp = do
-  -- Get the cabal directory from the cradle
   cradleConf <- BIOS.findCradle fp
   case cradleConf of
-    Just yaml -> BIOS.loadCradle yaml
-    Nothing -> BIOS.loadImplicitCradle fp
\ No newline at end of file
+    Just yaml -> fixCradle <$> BIOS.loadCradle yaml
+    Nothing   -> cabalHelperCradle fp
+
+-- | Check if the given Cradle is a stack cradle.
+-- This might be used to determine the GHC version to use on the project.
+-- If it is a stack-Cradle, we have to use `stack path --compiler-exe`
+-- otherwise we may ask `ghc` directly what version it is.
+isStackCradle :: Cradle -> Bool
+isStackCradle = (`elem` ["stack", "Cabal-Helper-Stack"])
+  . BIOS.actionName
+  . BIOS.cradleOptsProg
+
+-- | Finds a Cabal v2-project, Cabal v1-project or a Stack project
+-- relative to the given FilePath.
+-- Cabal v2-project and Stack have priority over Cabal v1-project.
+-- This entails that if a Cabal v1-project can be identified, it is
+-- first checked whether there are Stack projects or Cabal v2-projects
+-- before it is concluded that this is the project root.
+-- Cabal v2-projects and Stack projects are equally important.
+-- Due to the lack of user-input we have to guess which project it
+-- should rather be.
+-- This guessing has no guarantees and may change at any time.
+findCabalHelperEntryPoint :: FilePath -> IO (Maybe (Ex ProjLoc))
+findCabalHelperEntryPoint fp = do
+  projs <- concat <$> mapM findProjects subdirs
+  case filter (\p -> isCabalNewProject p || isStackProject p) projs of
+    (x:_) -> return $ Just x
+    []    -> case filter isCabalOldProject projs of
+      (x:_) -> return $ Just x
+      []    -> return Nothing
+
+    where
+      -- | Subdirectories of a given FilePath.
+      -- Directory closest to the FilePath `fp` is the head,
+      -- followed by one directory taken away.
+      subdirs :: [FilePath]
+      subdirs = reverse . map joinPath . tail . inits
+        $ splitDirectories (takeDirectory fp)
+
+      isStackProject (Ex ProjLocStackYaml {}) = True
+      isStackProject _ = False
+
+      isCabalNewProject (Ex ProjLocV2Dir {}) = True
+      isCabalNewProject (Ex ProjLocV2File {}) = True
+      isCabalNewProject _ = False
+
+      isCabalOldProject (Ex ProjLocV1Dir {}) = True
+      isCabalOldProject (Ex ProjLocV1CabalFile {}) = True
+      isCabalOldProject _ = False
+
+-- | Given a FilePath, find the Cradle the FilePath belongs to.
+--
+-- Finds the Cabal Package the FilePath is most likely a part of
+-- and creates a cradle whose root directory is the directory
+-- of the package the File belongs to.
+cabalHelperCradle :: FilePath -> IO Cradle
+cabalHelperCradle file = do
+  projM <- findCabalHelperEntryPoint file
+  case projM of
+    Nothing        -> do
+      errorm $ "Could not find a Project for file: " ++ file
+      error $ "Could not find a Project for file: " ++ file
```

Comment:
Is calling "error" the right thing? Probably should return the default cradle or none cradle.

Reply:
We can not continue to provide anything. The none cradle may still make sense!